core/src/main/java/io/questdb/cairo/TimestampDriver.java

```diff
@@ -0,0 +1,524 @@
+     * @param start     the start timestamp value
+     * @param end       the end timestamp value
+     * @param leftType  left column type
+     * @param rightType right column type
+     * @return the period between the timestamps in the specified unit
+     */
+    long getPeriodBetween(char unit, long start, long end, int leftType, int rightType);
+
+    /**
+     * Gets the quarter from a timestamp value.
+     *
+     * @param timestamp the timestamp value
+     * @return the quarter (1-4), or Numbers.INT_NULL if timestamp is null
+     */
+    int getQuarter(long timestamp);
+
+    /**
+     * Gets the second of minute from a timestamp value.
+     *
+     * @param timestamp the timestamp value
+     * @return the second of minute (0-59), or Numbers.INT_NULL if timestamp is null
+     */
+    int getSecondOfMinute(long timestamp);
+
+    int getTZRuleResolution();
+
+    long getTicks();
+
+    TimestampCeilMethod getTimestampCeilMethod(char c);
+
+    TimestampConstant getTimestampConstantNull();
+
+    TimestampDateFormatFactory getTimestampDateFormatFactory();
+
+    TimestampDiffMethod getTimestampDiffMethod(char type);
+
+    TimestampFloorMethod getTimestampFloorMethod(String c);
+
+    TimestampFloorWithOffsetMethod getTimestampFloorWithOffsetMethod(char c);
+
+    TimestampFloorWithStrideMethod getTimestampFloorWithStrideMethod(String c);
+
+    /**
+     * Creates a timestamp sampler instance for the given interval and time unit.
+     *
+     * @param interval the interval value
+     * @param timeUnit the time unit qualifier ('U', 'T', 's', 'm', 'h', 'd', 'w', 'M', 'y')
+     * @return a timestamp sampler instance
+     */
+    TimestampSampler getTimestampSampler(long interval, char timeUnit, int position) throws SqlException;
+
+    CommonUtils.TimestampUnitConverter getTimestampUnitConverter(int srcTimestampType);
+
+    TimeZoneRules getTimezoneRules(@NotNull DateLocale locale, @NotNull CharSequence timezone);
+
+    /**
+     * Gets the week of year from a timestamp value.
+     *
+     * @param timestamp the timestamp value
+     * @return the week of year, or Numbers.INT_NULL if timestamp is null
+     */
+    int getWeek(long timestamp);
+
+    /**
+     * Gets the year from a timestamp value.
+     *
+     * @param timestamp the timestamp value
+     * @return the year, or Numbers.INT_NULL if timestamp is null
+     */
+    int getYear(long timestamp);
+
+    default long implicitCast(CharSequence value, int typeFrom) {
+        assert typeFrom == ColumnType.STRING || typeFrom == ColumnType.SYMBOL;
+        if (value != null) {
+            try {
+                return Numbers.parseLong(value);
+            } catch (NumericException ignore) {
+            }
+
+            // Parse as ISO with variable length.
+            try {
+                return parseFloorLiteral(value);
+            } catch (NumericException ignore) {
+            }
+
+            return castPGDates(value, typeFrom, getColumnType());
```

Comment:
this looks suspicious - `castPGDates()` uses `to` and `from` types for reporting purposes only. this means it returns exactly the same value regardless of nanos/micros.

Reply:
tests added